server/src/main/java/org/opensearch/action/support/clustermanager/MasterThrottlingRetryListener.java

```diff
@@ -0,0 +1,131 @@
+ *
+ * It uses ExponentialEqualJitterBackoff policy for determining delay between retries.
+ */
+public class MasterThrottlingRetryListener<Request extends ClusterManagerNodeRequest<Request>, Response extends ActionResponse>
+    implements
+        ActionListener<Response> {
+
+    private static final Logger logger = LogManager.getLogger(MasterThrottlingRetryListener.class);
+
+    /**
+     * Base delay in millis.
+     */
+    private final int BASE_DELAY_MILLIS = 10;
+
+    /**
+     * Maximum delay in millis.
+     */
+    private final int MAX_DELAY_MILLIS = 5000;
+
+    private long totalDelay;
+    private final Iterator<TimeValue> backoffDelay;
+    private final ActionListener<Response> listener;
+    private final Request request;
+    private final Runnable runnable;
+    private final String actionName;
+    private final boolean localNodeRequest;
+
+    private static ScheduledThreadPoolExecutor scheduler = Scheduler.initScheduler(Settings.EMPTY);
+
+    public MasterThrottlingRetryListener(String actionName, Request request, Runnable runnable, ActionListener<Response> actionListener) {
+        this.actionName = actionName;
+        this.listener = actionListener;
+        this.request = request;
+        this.runnable = runnable;
+        this.backoffDelay = BackoffPolicy.exponentialEqualJitterBackoff(BASE_DELAY_MILLIS, MAX_DELAY_MILLIS).iterator();
+        /**
+         This is to determine whether request is generated from local node or from remote node.
+         If it is local node's request we need to perform the retries on this node.
+         If it is remote node's request, we will not perform retries on this node and let remote node perform the retries.
+
+         If request is from remote data node, then data node will set remoteRequest flag in {@link MasterNodeRequest}
+         and send request to master, using that on master node we can determine if the request was localRequest or remoteRequest.
+         */
+        this.localNodeRequest = !(request.isRemoteRequest());
+    }
+
+    @Override
+    public void onResponse(Response response) {
+        listener.onResponse(response);
+    }
+
+    @Override
+    public void onFailure(Exception e) {
+
+        if (localNodeRequest && isThrottlingException(e)) {
+            logger.info("Retrying [{}] on throttling exception from master. Error: [{}]", actionName, getExceptionMessage(e));
+            long delay = backoffDelay.next().getMillis();
+            if (totalDelay + delay >= request.clusterManagerNodeTimeout.getMillis()) {
+                delay = request.clusterManagerNodeTimeout.getMillis() - totalDelay;
+                scheduler.schedule(new Runnable() {
+                    @Override
+                    public void run() {
+                        listener.onFailure(new ProcessClusterEventTimeoutException(request.clusterManagerNodeTimeout, actionName));
+                    }
+                }, delay, TimeUnit.MILLISECONDS);
```

Comment:
no productive action is being done on a scheduler right?

Reply:
This class has been removed and reused the `RetryableAction` for retries.